pkg/agent/controller/networkpolicy/packetin.go

```diff
@@ -0,0 +1,176 @@
+	if pktIn == nil {
+		return errors.New("empty packetin for CNP")
+	}
+
+	ob := new(logInfo)
+
+	// Get network policy log info
+	err := getNetworkPolicyInfo(pktIn, c, ob)
+	if err != nil {
+		return err
+	}
+
+	// Get packet log info
+	err = getPacketInfo(pktIn, ob)
+	if err != nil {
+		return err
+	}
+
+	// Store log file
+	CNPLogger.Printf("%s %s %s %s SRC: %s DEST: %s %d %s", ob.tableName, ob.npName, ob.disposition, ob.ofPriority, ob.srcIP, ob.destIP, ob.pckLength, ob.protocolStr)
+	return nil
+}
+
+func getNetworkPolicyFullName(npName string, npNamespace string) string {
+	if npName == "" || npNamespace == "" {
+		return npName
+	} else {
+		return fmt.Sprintf("%s/%s", npNamespace, npName)
+	}
+}
+
+func getMatchRegField(matchers *ofctrl.Matchers, regNum uint32) *ofctrl.MatchField {
+	return matchers.GetMatchByName(fmt.Sprintf("NXM_NX_REG%d", regNum))
+}
+
+func getInfoInReg(regMatch *ofctrl.MatchField, rng *openflow13.NXRange) (uint32, error) {
+	regValue, ok := regMatch.GetValue().(*ofctrl.NXRegister)
+	if !ok {
+		return 0, errors.New("register value cannot be retrieved")
+	}
+	if rng != nil {
+		return ofctrl.GetUint32ValueWithRange(regValue.Data, rng), nil
+	}
+	return regValue.Data, nil
+}
+
+func getNetworkPolicyInfo(pktIn *ofctrl.PacketIn, c *Controller, ob *logInfo) error {
+	matchers := pktIn.GetMatches()
+	var match *ofctrl.MatchField
+	// Get table name
+	tableID := binding.TableIDType(pktIn.TableId)
+	ob.tableName = openflow.GetFlowTableName(tableID)
+
+	// Get ingress/egress reg
+	for _, table := range openflow.GetCNPEgressTables() {
+		if tableID == table {
+			match = getMatchRegField(matchers, uint32(openflow.EgressReg))
+		}
+	}
+	for _, table := range openflow.GetCNPIngressTables() {
+		if tableID == table {
+			match = getMatchRegField(matchers, uint32(openflow.IngressReg))
+		}
+	}
+
+	// Get network policy full name, CNP is not namespaced
+	info, err := getInfoInReg(match, nil)
+	if err != nil {
+		return err
+	}
+	npName, npNamespace := c.ofClient.GetPolicyFromConjunction(info)
+	ob.npName = getNetworkPolicyFullName(npName, npNamespace)
+
+	// Get OF priority of the conjunction
+	ob.ofPriority = c.ofClient.GetPriorityFromConjunction(info)
+
+	// Get disposition Allow or Drop
+	match = getMatchRegField(matchers, uint32(openflow.DispositionReg))
+	info, err = getInfoInReg(match, nil)
+	if err != nil {
+		return err
+	}
+	ob.disposition = openflow.DispositionToString[info]
+	return nil
+}
+
+func getPacketInfo(pktIn *ofctrl.PacketIn, ob *logInfo) error {
+	if pktIn.Data.Ethertype == opsv1alpha1.EtherTypeIPv4 {
+		ipPacket, ok := pktIn.Data.Data.(*protocol.IPv4)
+		if !ok {
+			return errors.New("invalid IPv4 packet")
+		}
+		// Get source destination IP and protocol
+		ob.srcIP = ipPacket.NWSrc.String()
+		ob.destIP = ipPacket.NWDst.String()
+		ob.pckLength = ipPacket.Length
+		ob.protocolStr = opsv1alpha1.ProtocolsToString[int32(ipPacket.Protocol)]
+	}
+	return nil
+}
```

Comment:
Could you add newline at end of file? https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline